articles/121_realtime_proposal.md

```diff
@@ -0,0 +1,230 @@
+
+Another major consideration when designing a real-time system is the choice of operating system.
+A true "hard" real-time operating system is fully deterministic.
+"Soft" real-time operating systems are designed to meet most deadlines deterministically, but may still have an element of nondeterminism.
+Most conventional user-facing operating systems are not designed with determinism or fairness as the first priority, often because they have to support graphical displays and interactivity.
+Here are some features critical to deterministic performance that are often exposed to the user in an RTOS, but not in a conventional OS (1).
+For more information about why these features matter, refer to the introductory article on real-time systems on design.ros2.org (2).
+
+- Task scheduling
+
+  - Expose thread priority and deterministic schedulers that incorporate priority
+  - Allow full thread preemption
+
+- Concurrency and synchronization primitives
+
+  - Floored semaphores (allow a higher priority thread to take control of locked semaphore)
+
+- Interrupt handling
+
+  - Disable interrupts
+  - Mask interrupts (temporarily suspend interrupts in a critical section)
+
+- Memory allocation
+
+Some real-time systems have a non real-time component, e.g. for graphical display or other user interaction.
+This would require an operating system that allows non real-time threads to be scheduled alongside real-time threads.
+Because of the extra nondeterminism introduced by this requirement, this kind of system would be better suited to a soft real-time constraint.
+
+Another important constraint is which processor architectures are supported by which operating systems, and which RTOS is an appropriate choice due to performance constraints of the architecture.
+
+## Goal for ROS 2
+
+In order to support real-time systems in ROS 2, we propose an iterative approach for identifying limitations in the API that force the code path into performance bottlenecks.
+
+Select two use cases with real-time requirements that represent two extremes of the space described by diagram (#), for example:
+
+- An STM32 embedded board running no operating system or a hard real-time OS with 1 kHZ update frequency with 3% allowable jitter.
+- An x86 computer running Xenomai or RT_PREEMPT that receives the sensor data over Ethernet or USB and processes the output in soft or firm real-time, but also runs separate non-RT processes for user interaction.
+
+Implement the system with ROS 2.
+
+Profile the performance of the system.
+
+Identify solutions to improve the performance of the system, e.g. by exposing more concurrency primitives through the appropriate abstraction.
+
+Adapt the ROS 2 API or underlying libraries to implement these solutions and to improve the experience of writing real-time code.
+
+Refactor the system for the new API if necessary and repeat.
```

Comment:
Are the last five sentences some kind of bullets? If yes, should the become a list and a sentence before like "Next steps"?